### ISSUE TITLE
High Class A power draw FW1.3

### ISSUE THREAD

**Issue by Th3B1gKahuna** [danak6jq/RAK3172#12]
Hi Dana,

After a long break, I just updated STM32CubeIDE t0 the latest and the WL SDK firmware to 1.3.  Well I broke all the rules and made too many changes at once :-)  I built your code "as is" and it worked.  However I noticed that the current draw was between 1-2mA.

In comparison,  I measured one of my old class A projects based on the 1.1 FW (your code was the base) and the current draw was about 16uA.

It almost looks like it's not getting into STOP2 mode when idle.  I do have LowPower_Disable set to 0.

Have you ever measured the current draw (Class A) on this code?  Would be nice to know if I simply have something wrong in my environment, or I just need to dig deeper.

BTW - with the latest STMCubeIDE toolchain, I initially got linker errors and warnings.  By adding the following lines to main.c I was able to fix the errors:

__attribute__((weak)) void _close(void){}
__attribute__((weak)) void _lseek(void){}
__attribute__((weak)) void _read(void){}
__attribute__((weak)) void _write(void){}

Mahalo!

Glenn

**Comment by DominusFulguris** [comment]
I'm fighting the exact same issue right now as well.

What I have seen is that the low power manager is holding the device in sleep rather than stop and it is the UART/vcom that is holding it up.

**Comment by danak6jq** [comment]
Which base board are you testing the RAK3172 on? A WisBlock Base Board? If so, is it the first-generation with a micro USB connector, or the 2nd gen one with a USB-C connector? There's a known issue with the 2nd gen base board; the 3v3 switched power rail uses a bipolar transistor which draws about 1-2mA when the 3v3 rail is switched-on. This is the switch controlled by IO2 on the Base Board. If you're leaving the switched 3v3 on, this is exactly what you'll see.

IO2 is connected to the STM32WLE5 GPIO PA8.

The original Base Board uses a micropower switch which doesn't have this problem.

Check that IO2/PA8 is turned-off (set to 0), at least when sleeping. If you're using the switched 3v3 rail to power an I/O module that generates interrupts while sleeping, you'll want to disable that interrupt as well, at least for testing.

Further information:

https://forum.rakwireless.com/t/wisbase-rak5005-vs-rak19007-on-state-quiescent-current-for-3v3-s/9551

**Comment by danak6jq** [comment]
> BTW - with the latest STMCubeIDE toolchain, I initially got linker errors and warnings. By adding the following lines to main.c I was able to fix the errors:
> 
> attribute((weak)) void _close(void){}
> attribute((weak)) void _lseek(void){}
> attribute((weak)) void _read(void){}
> attribute((weak)) void _write(void){}

Those are apparently a result of the switch to ARM gcc 11.3, and is well-addressed here:

https://stackoverflow.com/questions/73742774/gcc-arm-none-eabi-11-3-is-not-implemented-and-will-always-fail

I opted to do:

> TL;DR
> Having said that, there is actually a method to get rid of the warnings from libnosys permanently: Look for libnosys.a in your toolchain directory (in arm-none-abi case there will one for each architecture and float ABI), and delete .gnu.warning symbols from archives using:
> 
> arm-none-eabi-objcopy -w -R .gnu.warning.* libnosys.a
> 
> Note that this will irreversibly modify your toolchain files. Most likely in intended, and non-harmful way, but still. Use with caution.

**Comment by danak6jq** [comment]
Details on the Base Board issue:

https://forum.rakwireless.com/t/wisbase-rak5005-vs-rak19007-on-state-quiescent-current-for-3v3-s/9551

**Comment by Th3B1gKahuna** [comment]
Thank you both for the quick response.

I'm using the older 5005 Base Board with the micro-usb.  I'm glad you brought this power issue to my attention on the new Base Board.  I'm going to be ordering some soon.   I will look at the interaction of the UART on the sleep/STOP mode as a starting point. 

Regarding the linking issues on the new 11.3 toolchain, I'll probably just install an older version and avoid the added complexity until it's properly sorted out.  Thanks for the pointers on this.

**Comment by danak6jq** [comment]
> I will look at the interaction of the UART on the sleep/STOP mode as a starting point.

Presumably the LPUART? The regular UART doesn't work in STOP.  Do you have something producing frequent interrupts? The MCU could be in STOP but waking often. Or it might not be in stop at all... I'm pretty sure I checked STOP current for the RAK3172+RAK5005 with this FW1.3 build but I'll take another look.

**Comment by Th3B1gKahuna** [comment]
It's amazing you just mentioned that.  I just found this and change from UART to LPUART.  Other than the change to LPUART, my .ioc file and all other code is the same as your FW1.3 distribution here on GitHub.

So I think the change to LPUART is having an effect.  STOP mode may be working because now my JOINs do not work. :-(  Sheesh - this all worked with the 1.1 firmware.

I'll go measure the current draw to verify STOP mode.

**Comment by Th3B1gKahuna** [comment]
Looks like something is still fundamentally wrong.
Current draw alternates between 2.1mA and .91mA.

**Comment by danak6jq** [comment]
Let's see if I can take a measurement here this evening

**Comment by Th3B1gKahuna** [comment]
I found the power draw issue.  I'm now back down to about 16uA.  It turns out there were extra GPIO pins being setup in BSP_RAK5005_Init().  Looks like these pins were used for some relay support.  I didn't see the pins used in .ioc pinout view.
Just commented out the BSP_RAK5005_Init() in lora_app.c and all is good.  Thanks for all your support.  Glenn.

**Comment by danak6jq** [comment]
Ah yes, the relay control stuff I'd built before sharing this workspace - my sincere apologies. I'll remove that. I realize now you were seeing a varying idle current draw, which suggests a floating input. Good catch and thank you!

**Comment by danak6jq** [comment]
I've removed the relay control from the FW1.3 branch but now I am not sure how that was creating a current draw in sleep since the two GPIOs used were outputs. GPIO PA8 is the 3V3_S line, which, on the RAK5005/19007 baseboards, switches a 3v3 rail as needed. 

There is one input - PB12 - used to sense the RAK3172 band. It's tied high or low inside the RAK3172 module so should not be an issue for leakage, and it's initialized elsewhere (in the CubeMX framework).

Can you confirm that PA8 wasn't actually your issue? You probably want to initialize PA8 as an output and set it to 0 to keep the 3v3 rail turned-off when not in use.

**Comment by Th3B1gKahuna** [comment]
PA8 wasn’t the issue.  When set as an output:
State 0: Draw 16uA
State 1: Draw 22uA -> only a minor change when turning on the 3V3_S rail.

However Pin 11 and 12 did have an effect when they were initialized in BSP_RAK5005_Init().
Pin11: Accounted for 600uA draw
Pin12: Accounted for 300uA draw

Pin 11 and Pin12 were additive.  Not sure what’s going on here.

**Comment by danak6jq** [comment]
@Th3B1gKahuna fascinating. They're initialized as outputs. Do you have something connected to them? If they're NC, I can't imagine a current draw. Sounds like ~5k resistor connected to pin 11, 10k to pin 12. Pull-up/downs?

**Comment by danak6jq** [comment]
@Th3B1gKahuna I had better clarify. PB12 in the RAK3172 module is 10 kΩ internally pulled-up for high freq variant or pulled-down for low freq variant, but I initialize it as an output for another use as IO4 on the RAK5005. So depending what state it's in, there's a parasitic ~330uA load on it (a problem with using a GPIO as a boot-time input and run-time output). PA11 is initialized as an output for use as I2C_SDA, which might be pulled high via ~4k7 (~600uA) on the base board... hah!

Looking at the RAK5005 schematic, there is in fact a 4k7 pull-up on PA11 and PA12, which **are** used for I2C1.

So now we're understanding what those pins are doing. Not sure why you didn't see these loads before, I reckon they're always present. This isn't exactly a _bug_ but it's def in need of customization in your application.

Have a look at the RAK documentation for the RAK3172 module, the WisCore with it, and the RAK5005 and I think it all makes sense now.

**Comment by Th3B1gKahuna** [comment]
Nothing at all connected to the board.  It sure does sound like a 5K and 10K current draw.  I'm going to continue scouring the code to see if there's a hidden gem.  It would be nice to see exactly what’s inside the RAK3172 and how the STM32WL is used within.

**Comment by danak6jq** [comment]
The 10k pull-up on PB12 is inside the RAK3172 (see the RAK datasheet). The 4k7 pull-up on PA11 is on the RAK5005 (see the RAK datasheet). If you don't see them readily, I can paste clips from the schematic.

RAK isn't going to release the internal schematic of the RAK3172 but it's not complicated and you can suss it out from the datasheet (other than RF matching/switching bits). The RAK5005 schematic is in the the datasheet.

**Comment by Th3B1gKahuna** [comment]
Now that all makes sense.  I was able to find the resistors on the schematics.  It’s interesting how you don’t see or pay attention to these details until you're counting uA.
Really appreciate your help and insights.

Glenn. N6URO

**Comment by danak6jq** [comment]
@Th3B1gKahuna "It’s interesting how you don’t see or pay attention to these details until you're counting uA."

Heh! 100% my life the last year. 1M pull-ups are a 3uA load, 100k are 30uA. You can guess how I know this off the top of my head now!

73 OM

**Comment by DominusFulguris** [comment]
I have taken a different path on the problem after seeing some issues with migrating ioc files across to later versions of STM32CubeIDE (e.g. missing functions such as _write())

I now have a procedure that goes from "New Project" (STM32CubeIDE v1.13.1) to an operating end node without reference to outside code.  It is host board independent, focussing entirely on the RAK3172 and RAK3172-T. That is, doesn't know nor care about LEDs, buttons, etc.

If anyone wants to test the procedure out, DM me for a copy of the document.

**Comment by Th3B1gKahuna** [comment]
@DominusFulguris I'm interested to take a look.

**Comment by danak6jq** [comment]
I am also interested. There's quite a bit of detail in the IOC that varies from default, I took the relatively easy approach of just sharing my workspace. Would love to see this.

**Comment by danak6jq** [comment]
@DominusFulguris 
> I have taken a different path on the problem after seeing some issues with migrating ioc files across to later versions of STM32CubeIDE (e.g. missing functions such as _write())

BTW the _write() missing, nolib stuff, was the result ST updating to the GCC 11.x toolchain, not really an IOC/Cube thing.

**Comment by DominusFulguris** [comment]
> @DominusFulguris
> 
> > I have taken a different path on the problem after seeing some issues with migrating ioc files across to later versions of STM32CubeIDE (e.g. missing functions such as _write())
> 
> BTW the _write() missing, nolib stuff, was the result ST updating to the GCC 11.x toolchain, not really an IOC/Cube thing.

Oh, I understand that.

But generating the project from scratch also solves the problem without having to add in dummy functions.

**Comment by DominusFulguris** [comment]
Document link below.

There are quite a few steps and it is easy to miss some. The IOC widget also seems to reset parameters in one place after making changes in another place,  so it is worth giving the config a check over before generating code.

https://www.dropbox.com/scl/fi/go9zwlwshpinmquye9a5j/Project-Setup-for-RAK3172-module-in-STM32CubeIDE-v1.13.pdf?rlkey=g9w1ikr4pzvsff53w8j5ywou4&dl=0

**Comment by DominusFulguris** [comment]
[Project Setup for RAK3172 module in STM32CubeIDE v1.13.pdf](https://github.com/danak6jq/RAK3172/files/12494326/Project.Setup.for.RAK3172.module.in.STM32CubeIDE.v1.13.pdf)

There a quite a few steps so it is worth double checking everything.

**Comment by danak6jq** [comment]
> [Project Setup for RAK3172 module in STM32CubeIDE v1.13.pdf](https://github.com/danak6jq/RAK3172/files/12494326/Project.Setup.for.RAK3172.module.in.STM32CubeIDE.v1.13.pdf)
> 
> There a quite a few steps so it is worth double checking everything.

This is outstanding!

I see you cover the XO capacitance setting. Excellent!

**Comment by DominusFulguris** [comment]
Revision made to the section on radio_conf.h

[Project Setup for RAK3172 module in STM32CubeIDE v1.13 rev 1.2.pdf](https://github.com/danak6jq/RAK3172/files/12509811/Project.Setup.for.RAK3172.module.in.STM32CubeIDE.v1.13.rev.1.2.pdf)
